### PR TITLE
Initial v5.0.0 migration guide

### PR DESCRIPTION
**Pull request by agubler**
**Type:** feature

The following has been addressed in the PR:

* [ ] There is a related issue
* [ ] All code has been formatted with [`prettier`](https://prettier.io/) as per the [readme code style guidelines](./../#code-style)
* [ ] Unit or Functional tests are included in the PR

**Description:**

An initial v5.0.0 migration guide.

### CODE REVIEW

File: docs/V5-Migration-Guide.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
+# Version 5.0.0 Migration Guide
+
+Dojo 5.0.0 includes a few breaking changes that will need to some changes in your applications code base. As with previous releases, we have updated the `@dojo/cli-upgrade-app` CLI command to help you through the upgrade process and automate as much as possible. Where it is not possible to automate the upgrade, you should receive helpful information in the output that indicates what needs to get manually upgraded and in which module.
```
Comment (dylans) [inline docs/V5-Migration-Guide.md]:
```suggestion
Dojo 5.0.0 introduces a few breaking changes requiring changes in your application's code base. As with previous releases, we have updated the `@dojo/cli-upgrade-app` CLI command to help you through the upgrade process and automate the migration as much as possible. Where it is not possible to automate the upgrade, you should receive helpful information in the output that indicates what needs to get manually upgraded and in which module.
```

File: docs/V5-Migration-Guide.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+To install the upgrade command, run the following from the project root:
+
+```
+npm install @dojo/cli-upgrade-app --no-save
+```
+
+To perform the migration, run the following command from the project root. The migrations tool should automatically detect the necessary migration from your `package.json`.
```
Comment (dylans) [inline docs/V5-Migration-Guide.md]:
```suggestion
To perform the migration, run the following command from the project root. The migration tool should automatically detect the necessary migration from your `package.json`.
```

File: docs/V5-Migration-Guide.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+```
+dojo upgrade app
+```
+
+If you are upgrading from a version before 4.0.0, please see the [previous migration guides](./) for more details first.
+
+* The `@dojo/cli` should be updated to version 5.0.0, along with all the commands used by the project.
+* If your project is using @dojo/widgets and @dojo/interop, these packages also require upgrading to version 5.0.0.
```
Comment (dylans) [inline docs/V5-Migration-Guide.md]:
```suggestion
* If your project is using `@dojo/widgets` and `@dojo/interop`, these packages also require upgrading to version 5.0.0.
```

File: docs/V5-Migration-Guide.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+**Note:** The migration tool may create line lengths that violate your projects linting rules, be sure to run your linter and manually fix any linting rule violations.
+
+### Breaking Changes:
+
+#### [has consolidation](https://github.com/dojo/framework/pulls)
+
+Part of the legacy clean up, this release of Dojo consolidates all the `has` related modules in a single `has` module, `@dojo/framework/has/has`. If you run the `dojo/cli-upgrade-app` command all references to removed has modules should be automatically updated.
```
Comment (dylans) [inline docs/V5-Migration-Guide.md]: Will someone know the context for the legacy cleanup? Maybe say something like "Part of the cleanup started with the 3.0 consolidation of packages into `@dojo/framework` ?

Comment (dylans) [inline docs/V5-Migration-Guide.md]:
```suggestion
Part of the legacy clean up, this release of Dojo consolidates all the `has` related modules in a single `has` module, `@dojo/framework/has/has`. If you run the `dojo/cli-upgrade-app` command all references to removed `has` modules should be automatically updated.
```

File: docs/V5-Migration-Guide.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+#### [Support for `before` middleware in stores](https://github.com/dojo/framework/pulls)
+
+Middleware in stores has always support running after a process completion or error, Dojo 5.0.0 introduces support for middleware that runs before a process is started. To support this the API has changed. More information can be found on the [stores readme](https://github.com/dojo/framework/blob/master/src/stores/README.md#middleware).
```
Comment (dylans) [inline docs/V5-Migration-Guide.md]:
```suggestion
Middleware in stores has always support running after a process completion or error, Dojo 5.0.0 introduces support for middleware that runs before a process is started. The API has changed to support this middleware addition. More information can be found on the [stores readme](https://github.com/dojo/framework/blob/master/src/stores/README.md#middleware).
```

File: docs/V5-Migration-Guide.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+const myMiddlewareTwo = () => {
+	after: (error, result) => {
+		// middleware logic
+	}
+}
+
+const myProcess = createProcess('process', [myCommand], [myMiddlewareOne, myMiddlewareTwo]);
+```
+
+**Note:** If your application is using the `createCallbackDecorator` factory from `@dojo/framework/stores/process` these processes will be forwards compatible as the factory has been re-purposed to convert the 4.0.0 API to the 5.0.0 API. However we recommend updating these as time permits and using the `createCallbackDecorator` factory will log out a warning to the console when running a dev build.
```
Comment (dylans) [inline docs/V5-Migration-Guide.md]:
```suggestion
**Note:** If your application is using the `createCallbackDecorator` factory from `@dojo/framework/stores/process` these processes will be forwards-compatible as the factory has been re-purposed to convert the 4.0.0 API to the 5.0.0 API. However, we recommend updating these as time permits. Using the `createCallbackDecorator` factory will log a warning to the console when running a dev build.
```

**Comment by agubler** [comment]
@dylans Updated based on suggestions

**Pull request merged**